<a href="https://colab.research.google.com/github/hiephamk/Data_Analysis/blob/main/GroupHCDDFinal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Analysis & Visualization - Project Work (Team HCDA)

## Introduction & Objective

We selected dataset from Kaggle: https://www.kaggle.com/datasets/arnabchaki/data-science-salaries-2023 on 15 Nov 2023. The information this dataset was most likely gathered by scraping job postings from ai-jobs.net.

We aim to use this dataset to analyze salaries for various data science roles, factoring in key variables like experience, location, and job specifics. This offers insights into potential earnings across different domains that influence data science salaries. For example, it can be used to address the following questions:

1. Which are the top 10 jobs with the highest salary for juniors in 2023?
2. What is the mean annual salary of each job title in small-sized companies in the US?
3. How does the average income for the Analytics Lead job vary by experience level?

For those embarking on the data science journey, these observations are like having a crystal ball. It enables informed decision-making and strategic job hunting, thereby setting realistic salary expectations.

# 1. Data Preprocessing

**Activities:**
- Select a dataset to work on and understand your dataset. Your activities may include tasks such as viewing a random sample of data, getting the total number of rows and columns.
- Check to see if your dataset contains any missing values and get the percentage of the missing data. Within the context of your dataset, decide what to do with the missing values and take necessary steps.
Identify and drop duplicate values from the dataset.
- Separate one of your columns in the dataset that contains continuous numeric data into appropriate bins. You may use cut or qcut function.
- Identify any outliers within your dataset. If the dataset does not include any outlier, you can randomly mess up some portion of your data.
Decide what to do with the outliers.

**Assessment:**
- Understanding the Dataset (1 point): Demonstrates a basic understanding of the selected dataset by performing tasks like viewing a random sample, determining the total number of rows and columns.
- Handling Missing Values (1 point): Identifies and addresses missing values effectively, providing a clear strategy for handling them.
- Handling Duplicates (1 point): Detects and removes duplicate values from the dataset, ensuring data cleanliness.

**Notes to get teacher's advise**
- What should we do if there is no missing values in our data set?
- Same with duplicated values
- There is no binary value either, can we make assumption on a specific column? Or should we run the command & show the null result?

### Data Exploration

In [ ]:
# Import neccessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')
warnings.simplefilter('ignore', DeprecationWarning)

In [ ]:
# Load raw data
from google.colab import data_table
url = 'https://raw.githubusercontent.com/hiephamk/Data_Analysis/main/ds_salaries.csv'
salary = pd.read_csv(url,sep = ',')

In [ ]:
# Preview the first 10 lines of the loaded data
salary.head(10)

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   
5       2023               SE              FT         Applied Scientist   
6       2023               SE              FT         Applied Scientist   
7       2023               SE              FT            Data Scientist   
8       2023               SE              FT            Data Scientist   
9       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                 ES           100   
1   30000             USD          30000                 US           100   
2   25500             USD          25500                 US           100   
3  175000             USD         175000                 CA           100   
4  120000             USD         120000                 CA           100   
5  222200             USD         222200                 US             0   
6  136000             USD         136000                 US             0   
7  219000             USD         219000                 CA             0   
8  141000             USD         141000                 CA             0   
9  147100             USD         147100                 US             0   

  company_location company_size  
0               ES            L  
1               US            S  
2               US            S  
3               CA            M  
4               CA            M  
5               US            L  
6               US            L  
7               CA            M  
8               CA            M  
9               US            M

In [ ]:
# View a random dataset of data
sample = salary.sample(5)
sample

work_year experience_level employment_type                  job_title  \
1415       2023               EN              FT              Data Engineer   
3539       2021               MI              FT     Applied Data Scientist   
212        2023               SE              FT  Machine Learning Engineer   
2210       2022               SE              FT              Data Engineer   
2662       2022               MI              FT             Data Scientist   

      salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
1415  160000             USD         160000                 US             0   
3539   68000             CAD          54238                 GB            50   
212   172200             USD         172200                 US             0   
2210   95000             USD          95000                 US             0   
2662  108000             USD         108000                 US            50   

     company_location company_size  
1415               US            M  
3539               CA            L  
212                US            M  
2210               US            M  
2662               US            L

Now we need to know how many rows and columns are there in the data set, what variables it has and the data types of the variables.

In [ ]:
# Viewing data type and getting the total number of rows and columns
print(salary.info())
salary.shape

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3755 entries, 0 to 3754
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           3755 non-null   int64 
 1   experience_level    3755 non-null   object
 2   employment_type     3755 non-null   object
 3   job_title           3755 non-null   object
 4   salary              3755 non-null   int64 
 5   salary_currency     3755 non-null   object
 6   salary_in_usd       3755 non-null   int64 
 7   employee_residence  3755 non-null   object
 8   remote_ratio        3755 non-null   int64 
 9   company_location    3755 non-null   object
 10  company_size        3755 non-null   object
dtypes: int64(4), object(7)
memory usage: 322.8+ KB
None


(3755, 11)

As we can see from the above data frame information, there are 3755 rows, 11 columns, and no missing values in this dataset since all fields has 3755 non-null entries.

However, we could check again to assure if there are some missing values in the data set.

In [ ]:
# Check missing values
missing_values = salary.isnull().sum()
print(missing_values)

# Calculate the missing percentage
total_values = np.product(salary.shape)
percentage = (missing_values / total_values) * 100
print(percentage)

work_year             0
experience_level      0
employment_type       0
job_title             0
salary                0
salary_currency       0
salary_in_usd         0
employee_residence    0
remote_ratio          0
company_location      0
company_size          0
dtype: int64
work_year             0.0
experience_level      0.0
employment_type       0.0
job_title             0.0
salary                0.0
salary_currency       0.0
salary_in_usd         0.0
employee_residence    0.0
remote_ratio          0.0
company_location      0.0
company_size          0.0
dtype: float64


As per the above result, we now can ensure that our data set has no missing value.

### Data Transformation

#### Unnecessary data drop

Firstly, out of 3,755 observations in the dataset, there are only over 700 data values from other nations besides the US. Since this data may not be meaningfully contributing to useful analysis, we decide to exclude these data and retain only the data values from the US.

In [ ]:
# Filter data values to keep the US country only
df_salary1 = salary[salary['company_location'] == 'US']
df_salary1.shape

(3040, 11)

Now that we only have one country in the data, we do not need 2 features "company_location" and "employee_residence" because all employees who work for a company based in the US also live in the US. So we will drop them for a cleaner data set.

Besides, the "salary" and "salary_currency" features will also be removed. We saw that there are 3 features relevant to salary: "salary", "salary_currency", and "salary_in_usa". However, we will primarily focus on the "salary_in_usd" feature because it provides a standardized currency for easy comparisons.

In [ ]:
# Drop 2 columns 'company_location' and 'employee_residence'
df_salary2 = df_salary1.drop(columns=['company_location', 'employee_residence'])
df_salary2.shape

(3040, 9)

In [ ]:
# Drop 2 columns: 'salary' and 'salary_currency'
df_salary3 = df_salary2.drop(['salary','salary_currency'], axis=1)
df_salary3.head()

work_year experience_level employment_type          job_title  \
1       2023               MI              CT        ML Engineer   
2       2023               MI              CT        ML Engineer   
5       2023               SE              FT  Applied Scientist   
6       2023               SE              FT  Applied Scientist   
9       2023               SE              FT     Data Scientist   

   salary_in_usd  remote_ratio company_size  
1          30000           100            S  
2          25500           100            S  
5         222200             0            L  
6         136000             0            L  
9         147100             0            M

In [ ]:
# Check the number of rows and columns of the latest dataset
df_salary3.shape

(3040, 7)

As observed, there are limited number of data values for the year 2020, 2021 (which are only account for over 130 out of 3,040 observations) in the latest data set. These data may not contribute significantly to our analytic, so we will exclude it in the final data set.

In [ ]:
# Filter data without values of years 2020 and 2021
df_salary = df_salary3[(df_salary3['work_year'] != 2020) & (df_salary3['work_year'] != 2021)]
df_salary.shape

(2909, 7)

Finally, our data set now has 2909 rows and 7 columns in total.

#### Duplicated values

We are going to identify and remove duplicated values (if any) in order to smooth our data set.

In [ ]:
# Check duplicate
df_salary.duplicated()

1       False
2       False
5       False
6       False
9       False
        ...  
3479    False
3482    False
3486    False
3490    False
3493    False
Length: 2909, dtype: bool

#### Distinct values

In [ ]:
unique_counts = df_salary.nunique(axis=0)

def valuesGet():
    unique_values_list = []
    for col_name in unique_counts.index:
        values_get = df_salary[col_name].unique()
        unique_values_list.append(values_get)
    return unique_values_list

unique_values = valuesGet()

df2 = pd.DataFrame({'Values_range': unique_values})

df1 = pd.DataFrame({'Column_name': unique_counts.index, 'Distinct_values_count': unique_counts.values})
unique_show =pd.concat([df1,df2],axis=1)
unique_show


Column_name  Distinct_values_count  \
0         work_year                      2   
1  experience_level                      4   
2   employment_type                      4   
3         job_title                     61   
4     salary_in_usd                    645   
5      remote_ratio                      3   
6      company_size                      3   

                                        Values_range  
0                                       [2023, 2022]  
1                                   [MI, SE, EN, EX]  
2                                   [CT, FT, PT, FL]  
3  [ML Engineer, Applied Scientist, Data Scientis...  
4  [30000, 25500, 222200, 136000, 147100, 90700, ...  
5                                       [100, 0, 50]  
6                                          [S, L, M]

#### Replace values
According to the above distinct values and source inspection, we can decribe the distinct values of several fields as following:

**1. Experience Level:** 4 unique values
- EN: Entry-level / Junior
- MI: Mid-level / Intermediate
- SE: Senior-level / Expert
- EX: Executive-level / Director

**2. Employee Type:** 4 categories
- PT: Part-time
- FT: Full-time
- CT: Contract
- FL: Freelance

**3. Remote Ratio:** 3 numerical values
- 0: Office work
- 50: Partial remote
- 100: Fully remote

**4. Company Size:** 3 unique values
- S: Small-sized company
- M: Medium-sized company
- L: Large-sized company

Now let's replace these abbreviations with the descriptions for easy understanding.

In [ ]:
# Replace the abbreviations
df_salary['experience_level'] = df_salary['experience_level'].replace(['EN','MI','SE','EX'], ['Entry-level','Mid-level','Senior','Executive'])
df_salary['employment_type'] = df_salary['employment_type'].replace(['FT','PT','CT','FL'], ['Full-time','Part-time','Contract','Freelance'])
df_salary['remote_ratio'] = df_salary['remote_ratio'].replace([0,50,100], ['Office work','Partial remote','Fully remote'])
df_salary['company_size'] = df_salary['company_size'].replace(['S', 'M', 'L'], ['Small', 'Medium', 'Large'])

In [ ]:
# Show 5 first rows of the final data set
df_salary.head()

work_year experience_level employment_type          job_title  \
1       2023        Mid-level        Contract        ML Engineer   
2       2023        Mid-level        Contract        ML Engineer   
5       2023           Senior       Full-time  Applied Scientist   
6       2023           Senior       Full-time  Applied Scientist   
9       2023           Senior       Full-time     Data Scientist   

   salary_in_usd  remote_ratio company_size  
1          30000  Fully remote        Small  
2          25500  Fully remote        Small  
5         222200   Office work        Large  
6         136000   Office work        Large  
9         147100   Office work       Medium

#### Discretization & Bining

In [ ]:
# Binning cut the salary range
salary_range = df_salary.salary_in_usd
salary_range

1        30000
2        25500
5       222200
6       136000
9       147100
         ...  
3479    100000
3482    144000
3486    100000
3490    120000
3493    125000
Name: salary_in_usd, Length: 2909, dtype: int64

In [ ]:
salary_min = df_salary['salary_in_usd'].min()
salary_min

12000

In [ ]:
salary_max = df_salary['salary_in_usd'].max()
salary_max

405000

In [ ]:
# we will use cut function
bins = [12000,30000,50000,100000,150000,200000,410000]
salary_range_bins = pd.cut(salary_range,bins)
salary_range_bins

1         (12000, 30000]
2         (12000, 30000]
5       (200000, 410000]
6       (100000, 150000]
9       (100000, 150000]
              ...       
3479     (50000, 100000]
3482    (100000, 150000]
3486     (50000, 100000]
3490    (100000, 150000]
3493    (100000, 150000]
Name: salary_in_usd, Length: 2909, dtype: category
Categories (6, interval[int64, right]): [(12000, 30000] < (30000, 50000] < (50000, 100000] <
                                         (100000, 150000] < (150000, 200000] < (200000, 410000]]

In [ ]:
#square brakcet means it is closed and inclusive
#possible to change which side is closed by passing right=False
salary_range_bins2 = pd.cut(salary_range,bins, right=False)
salary_range_bins2

1         [30000, 50000)
2         [12000, 30000)
5       [200000, 410000)
6       [100000, 150000)
9       [100000, 150000)
              ...       
3479    [100000, 150000)
3482    [100000, 150000)
3486    [100000, 150000)
3490    [100000, 150000)
3493    [100000, 150000)
Name: salary_in_usd, Length: 2909, dtype: category
Categories (6, interval[int64, left]): [[12000, 30000) < [30000, 50000) < [50000, 100000) <
                                        [100000, 150000) < [150000, 200000) < [200000, 410000)]

In [ ]:
#lets count the values for each bin
pd.value_counts(salary_range_bins)

(100000, 150000]    1121
(150000, 200000]     805
(200000, 410000]     492
(50000, 100000]      459
(30000, 50000]        23
(12000, 30000]         8
Name: salary_in_usd, dtype: int64

In [ ]:
#lets count the values for each bin
pd.value_counts(salary_range_bins2)

[100000, 150000)    1119
[150000, 200000)     840
[200000, 410000)     545
[50000, 100000)      383
[30000, 50000)        15
[12000, 30000)         7
Name: salary_in_usd, dtype: int64

#2. Data Objects & Attribute Types

**Activities:**
- Use your dataset and print columns name that represent nominal attributes.
- Use your dataset and print columns name that represent binary attributes.
- Use your dataset and print columns name that represent ordinal attributes.

**Assessment:**
- Nominal Attributes (0.5 points): Accurately identifies and prints the column names representing nominal attributes.
- Binary Attributes (0.25 points): Accurately identifies and prints the column names representing binary attributes.
- Ordinal Attributes (0.25 points): Accurately identifies and prints the column names representing ordinal attributes.

###  Create a medthod to print data_attributes as a list

In [ ]:
def Data_print(input_data):
  for i, col in enumerate(input_data):
    print(i,col)

### Printing columns name that represent nominal attributes.

In [ ]:
print("Nominal Atribute columns:\n")
Nominal_col = df_salary.loc[:, ~df_salary.columns.isin(['salary_in_usd', 'company_size', 'experience_level'])]
Data_print(Nominal_col)

Nominal Atribute columns:

0 work_year
1 employment_type
2 job_title
3 remote_ratio


###  Printing columns name that represent binary attributes.

In [ ]:
Binary_col = salary.columns[salary.isin([0, 1]).all()]

if not Binary_col.empty:
    print("Binary Atribute columns:\n")
    for i, col in enumerate(Binary_col):
        print(i, col)
else:
    print("There are no binary columns in the data frame.")

There are no binary columns in the data frame.


### Printing columns name that represent ordinal attributes.

In [ ]:
print("Ordinal Atribute columns:\n")
Ordinal_col = df_salary[['experience_level','company_size']]
Data_print(Ordinal_col)

Ordinal Atribute columns:

0 experience_level
1 company_size


###  Printing columns name that represent numerical attributes.

In [ ]:
print("Numerical Atribute columns:\n")
Numerical_columns = df_salary[['salary_in_usd']]
Data_print(Numerical_columns)


Numerical Atribute columns:

0 salary_in_usd


In [ ]:
Numerical_columns.describe()

salary_in_usd
count    2909.000000
mean   152500.606050
std     54658.328851
min     12000.000000
25%    115000.000000
50%    145000.000000
75%    185000.000000
max    405000.000000

#3. Basic Statistics

**Activities:**
In this task, you will utilize NumPy to perform fundamental statistical operations on your dataset. You can explore the following calculations as a guideline:
- Mean (Calculate the mean or average )
- Median (Find the median value)
- Standard Deviation (Compute the standard deviation of data).
- Variance (Calculate the variance of data)
- Minimum and Maximum (Find the minimum and maximum values)
- Sum and Product (Compute the sum and product)

**Assessment:**
- Mean and Median (0.5 points): Correctly calculates the mean and median values for appropriate columns.
- Standard Deviation and Variance (0.5 points): Accurately computes the standard deviation and variance of data.
- Minimum and Maximum (0.25 points): Finds the minimum and maximum values.
- Sum and Product (0.25 points): Accurately calculates the sum and product of data.

Based on the sample dataset mimicking U.S.-based employees in 2023 with an 'EN' experience level, here are the basic statistical calculations:

**1. Mean (Average) Salary:** The mean salary is approximately \$126,845.18. This is the average salary across all the data points.

**2. Median Salary:** The median salary is \$130,971.50. This value represents the middle salary in the dataset when the values are sorted in ascending order. It is a useful measure of central tendency, especially in datasets with outliers.

**3. Standard Deviation:** The standard deviation of the salary data is approximately \$41,009.81. This measures how spread out the salaries are from their mean. A higher standard deviation indicates more variability in salary values.

**4. Variance:** The variance of the salary data is approximately 1,681,804,382.71. Variance quantifies the degree of spread in the salary data squared.

**5. Minimum and Maximum Salary:** The minimum salary in the dataset is \$53,918, and the maximum salary is \$198,663. These values show the range of salaries.

**6. Sum of Salaries:** The total sum of all salaries in the dataset is \$12,684,518. This gives an idea of the total salary expenditure for the group.

**7. Product of Salaries:** The product of all salary values is 0. This is typically not a meaningful statistic for salary data, and due to the large range of salary values, the product results in an overflow leading to a value of 0.

It's important to note that these calculations are based on a simulated dataset and should be performed on the actual data for accurate insights.

#4. Data Visualization

**Activities:**
You will utilize Matplotlib to create a diverse range of plots and charts. Depending on the dataset and your analysis, you may generate the following visuals as needed:
- Line Plot
- Scatter Plot
- Bar Chart
- Histrogram
- Piechart
- Heatmap
- 3D Plot

**Assessment:**
Creates both a line plot and a scatter plot with appropriate labels and titles.
- Bar Chart and Histogram (0.5 points): Successfully generates a bar chart and a histogram.
- Pie Chart and Heatmap (0.25 points): Produces a pie chart and a heatmap.
- 3D Plot (0.25 points): Creates a 3D plot if applicable to the dataset.